research/ag_news/sparse_tensor_exchange/server.py

```diff
@@ -0,0 +1,115 @@
+import argparse
+from functools import partial
+from logging import INFO
+from typing import Any, Dict
+
+import flwr as fl
+from flwr.common.logger import log
+from flwr.common.typing import Config
+from transformers import BertForSequenceClassification
+
+from fl4health.client_managers.poisson_sampling_manager import PoissonSamplingClientManager
+from fl4health.server.base_server import FlServer
+from fl4health.strategies.fedavg_sparse_coo_tensor import FedAvgSparseCooTensor
+from fl4health.utils.config import load_config
+from fl4health.utils.functions import get_all_model_parameters
+from fl4health.utils.metric_aggregation import evaluate_metrics_aggregation_fn, fit_metrics_aggregation_fn
+
+
+def construct_config(
+    current_round: int,
+    local_steps: int,
+    batch_size: int,
+    num_classes: int,
+    sample_percentage: float,
+    beta: float,
+) -> Config:
+    assert 0 < sample_percentage <= 1
+    assert 0 < beta
+    return {
+        "current_server_round": current_round,
+        "local_steps": local_steps,
+        "batch_size": batch_size,
+        "num_classes": num_classes,
+        "sample_percentage": sample_percentage,
+        "beta": beta,
+    }
+
+
+def fit_config(
+    local_steps: int,
+    batch_size: int,
+    num_classes: int,
+    sample_percentage: float,
+    beta: float,
+    current_round: int,
+) -> Config:
+    return construct_config(
+        current_round,
+        local_steps,
+        batch_size,
+        num_classes,
+        sample_percentage,
+        beta,
+    )
+
+
+def main(config: Dict[str, Any], server_address: str) -> None:
+    # This function will be used to produce a config that is sent to each client to initialize their own environment
+    fit_config_fn = partial(
+        fit_config,
+        config["local_steps"],
+        config["batch_size"],
+        config["num_classes"],
+        config["sample_percentage"],
+        config["beta"],
+    )
+
+    initial_model = BertForSequenceClassification.from_pretrained(
+        "google-bert/bert-base-cased", num_labels=config["num_classes"]
+    )
+
+    # Since clients can send different tensors to the server, we perform weighted averaging separately on each tensor.
+    strategy = FedAvgSparseCooTensor(
+        # Server waits for min_available_clients before starting FL rounds
+        min_available_clients=config["n_clients"],
+        on_fit_config_fn=fit_config_fn,
+        on_evaluate_config_fn=fit_config_fn,
+        fit_metrics_aggregation_fn=fit_metrics_aggregation_fn,
+        evaluate_metrics_aggregation_fn=evaluate_metrics_aggregation_fn,
+        initial_parameters=get_all_model_parameters(initial_model),
+    )
+
+    client_manager = PoissonSamplingClientManager()
```

Comment:
Given that we're not subsampling the clients, I don't think this matters, but you can likely just use a `SimpleClientManager` here.

Reply:
Yes good point